### ISSUE TITLE
Crashes with FITSCustomOptic unless size and pixel scale are identical

### ISSUE THREAD

**Issue by mperrin** [spacetelescope/poppy#207]
<a href="https://github.com/mb2448"><img src="https://avatars3.githubusercontent.com/u/1314227?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [mb2448](https://github.com/mb2448)**
_Friday Mar 10, 2017 at 19:44 GMT_
_Originally opened as https://github.com/mperrin/poppy/issues/207_

----

I'm trying to simulate Fresnel propagation with a custom optical element.  For the example below, it's a square stop.  The real one is not analytic and more complex.

The following code crashes due to an index error when poppy tries to rescale the array.

```
import matplotlib.pyplot as plt
import poppy
import astropy.units as u
import numpy as np
import pyfits as pf
import logging
logging.basicConfig()

def write_custom_optic(fname = 'custom.fits', size = 1024):
    """Defines a square stop"""
    q = np.ones((size, size))
    cent = size//2
    q[cent-100:cent+100, cent-100:cent+100] = 0

    #write to fits file since that's what FITSOpticalElement accepts
    hdu = pf.PrimaryHDU(q, header = None)
    hdulist = pf.HDUList([hdu])
    #hdulist[0].header.set('PUPLSCAL', .00001)
    hdulist.writeto(fname,clobber = True)
    hdulist.close()
    return fname
    

if __name__ == "__main__":
    npix = 1024 
    wf = poppy.FresnelWavefront(25.4e-3*u.m,wavelength=675e-9,npix=npix,oversample=1)
    wf *= poppy.CircularAperture(radius=25.4e-3) 
    wf *= poppy.GaussianAperture(name = 'gauss', fwhm = 25.4e-3*u.m) 
    
    filename = write_custom_optic(fname = 'custom.fits') 
    #changing the following to pixelscale = wf.pixelscale.value works
    myoptic = poppy.FITSOpticalElement(transmission = filename, pixelscale = 1e-5)
    wf *= myoptic
    wf.propagate_fresnel(1.0*u.m)
```

(The error message is reproduced below.)

Defining 'PUPLSCAL' in the fits header still causes the same error.  A workaround can be had when both the size and the pixel scale are the same between the wavefront and custom optic, but this involves defining a pixel scale on the wavefront and zero padding until everything is matched.  For example, replacing the `myoptic` line with

`myoptic = poppy.FITSOpticalElement(transmission = filename, pixelscale = wf.pixelscale.value)
`
runs just fine.

As a separate note, the documentation here:
_https://pythonhosted.org/poppy/extending.html_
says that `pupilscale` is an acceptable keyword to pass to `FITSOpticalElement`.  It doesn't accept that anymore.

Best regards,
MB

> /Users/Me/Desktop/Coronagraphy/poppy/poppy_practice.py in <module>()
>      31     #changing the following to pixelscale = wf.pixelscale.value works
>      32     myoptic = poppy.FITSOpticalElement(transmission = filename, pixelscale = 1e-5)
> ---> 33     wf *= myoptic
>      34     wf.propagate_fresnel(1.0*u.m)
> 
> /usr/local/lib/python2.7/site-packages/poppy/fresnel.pyc in __imul__(self, optic)
>     817         else:
>     818             # Otherwise fall back to the parent class
> --> 819             return super(FresnelWavefront, self).__imul__(optic)
>     820 
>     821     def apply_lens_power(self, optic, ignore_wavefront=False):
> 
> /usr/local/lib/python2.7/site-packages/poppy/poppy_core.pyc in __imul__(self, optic)
>     193             return self
>     194 
> --> 195         phasor = optic.get_phasor(self)
>     196 
>     197         if not np.isscalar(phasor) and phasor.size > 1:
> 
> /usr/local/lib/python2.7/site-packages/poppy/poppy_core.pyc in get_phasor(self, wave)
>    2283                     self._resampled_amplitude = np.zeros([lx_w,ly_w])
>    2284 
> -> 2285                     self._resampled_opd[border_x:border_x+resampled_opd.shape[0],border_y:border_y+resampled_opd.shape[1]] = resampled_opd
>    2286                     self._resampled_amplitude[border_x:border_x+resampled_opd.shape[0],border_y:border_y+resampled_opd.shape[1]]=resampled_amplitude
>    2287                     _log.debug("padded an optic with a %i x %i border to optic to match the wavefront"%(border_x,border_y))
> 
> TypeError: slice indices must be integers or None or have an __index__ method
>

**Comment by mperrin** [comment]
<a href="https://github.com/douglase"><img src="https://avatars1.githubusercontent.com/u/1025951?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [douglase](https://github.com/douglase)**
_Friday Mar 10, 2017 at 23:32 GMT_

----

@mb2448, I am having trouble reproducing this error, which versions of numpy and poppy are you using?  If you are running an older version of poppy (installed via pip for example) and numpy 1.12 could you be running into the indexing issue where numpy no-longer allows use of non-explicit integers, discussed in #203, which is fixed on the current master branch.

**Comment by mperrin** [comment]
<a href="https://github.com/mb2448"><img src="https://avatars3.githubusercontent.com/u/1314227?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mb2448](https://github.com/mb2448)**
_Friday Mar 10, 2017 at 23:51 GMT_

----

@douglase , I am using numpy 1.12 and poppy 0.5.1--is there a newer poppy version?

**Comment by mperrin** [comment]
<a href="https://github.com/douglase"><img src="https://avatars1.githubusercontent.com/u/1025951?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [douglase](https://github.com/douglase)**
_Saturday Mar 11, 2017 at 00:17 GMT_

----

A new version has not been "released" but 0.5.1 does not seem to be compatible with numpy 1.12 so you probably need to downgrade numpy or upgrade to the development version by cloning this repository (detailed instructions for installing development version: https://pythonhosted.org/poppy/installation.html).

@mperrin,@josephoenix, this issue will probably keep recurring, can we patch 0.5.1 to prevent installation with numpy 1.12 or is there a set of milestones you have in mind that I can help hit to release 0.6 in the near future?

**Comment by mperrin** [comment]
<a href="https://github.com/mperrin"><img src="https://avatars0.githubusercontent.com/u/1151745?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mperrin](https://github.com/mperrin)**
_Saturday Mar 11, 2017 at 00:30 GMT_

----

There's no holdup in releasing poppy 0.6, not apart from lack of time. It's more complex for webbpsf 0.6, but it's only for historical reasons and convenience that we've done simultaneous releases in the past.

**Comment by mperrin** [comment]
<a href="https://github.com/mperrin"><img src="https://avatars0.githubusercontent.com/u/1151745?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mperrin](https://github.com/mperrin)**
_Monday Aug 28, 2017 at 16:21 GMT_

----

I wanted to follow up on this old issue from a while ago. If you have time @mb2448 could you let me know if this problem was resolved, either by updating to poppy 0.6 or by some other fix? If there is still a bug we can take a shot at fixing it, but if it's working OK now I will simply close this issue. Thanks much!

**Comment by mperrin** [comment]
<a href="https://github.com/mperrin"><img src="https://avatars0.githubusercontent.com/u/1151745?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mperrin](https://github.com/mperrin)**
_Friday Apr 27, 2018 at 21:48 GMT_

----

This is an old/stale issue related to incompatibilities in older releases of both this and numpy, from a year ago. I don't believe any action is needed now, and there was no response to my query 6 months ago. so it seems safe to close this issue.
